include/oneapi/dpl/internal/distributed_ranges_impl/shp/algorithms/copy.hpp

```diff
@@ -0,0 +1,236 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <memory>
+#include <type_traits>
+
+#include <sycl/sycl.hpp>
+
+#include <oneapi/dpl/internal/distributed_ranges_impl/concepts/concepts.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/detail/segments_tools.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/shp/detail.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/shp/device_ptr.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/shp/util.hpp>
+
+namespace oneapi::dpl::experimental::dr::shp
+{
+
+// Copy between contiguous ranges
+template <std::contiguous_iterator InputIt, std::contiguous_iterator OutputIt>
+requires __detail::is_syclmemcopyable<std::iter_value_t<InputIt>, std::iter_value_t<OutputIt>> sycl::event
+copy_async(InputIt first, InputIt last, OutputIt d_first)
+{
+    auto&& q = __detail::get_queue_for_pointers(first, d_first);
+    return q.memcpy(std::to_address(d_first), std::to_address(first),
+                    sizeof(std::iter_value_t<InputIt>) * (last - first));
```

Comment:
Note (not for this PR approval, but for future work):

For something like this, I think eventually we will want this level of implementation to live within oneDPL main (dpcpp sycl backend) implementation of a `copy_async` routine.

Of course, I understand why this PR does not attempt this currently but it is something we need to consider as we look toward eventually coming out of experimental feature that we should try to share any lower level advantages, specializations, code improvements with oneDPL in general, and have distributed ranges operate at a higher level on top of those building blocks which are provided by oneDPL as it exists today.  

In other words, part of the work of maturing distributed ranges will need to be to contribute back any performance improvements into the rest of oneDPL to be usable without distributed ranges as well.

Reply:
Excellent point. We should make any changes necessary so we can just directly use oneDPL `copy` here as well as oneDPL's `for_each` instead of our workaround.